Step generator panics

This happens when I updated the username of a GKE cluster.

```
$ p up
warning: A new version of Pulumi is available. To upgrade from version '0.16.3-dev-1541007180-gb7489357-dirty' to '0.16.4', visit https://pulumi.io/install for manual instructions and release notes.
Previewing update (istio):

     Type                            Name         Plan       Info
     pulumi:pulumi:Stack             istio-istio
 ~   ├─ gcp:container:Cluster        gke-cluster  update     [diff: ~masterAuth]
 ~   └─ pulumi:providers:kubernetes  gkeK8s       update..   [diff: ~kubeconfig]
panic: fatal: An assertion has failed

goroutine 163 [running]:
github.com/pulumi/pulumi/pkg/util/contract.failfast(0x1abcd20, 0x17)
	/Users/alex/go/src/github.com/pulumi/pulumi/pkg/util/contract/failfast.go:23 +0xe9
github.com/pulumi/pulumi/pkg/util/contract.Assert(0xc00182d100)
	/Users/alex/go/src/github.com/pulumi/pulumi/pkg/util/contract/assert.go:26 +0x49
github.com/pulumi/pulumi/pkg/resource/plugin.(*provider).Diff(0xc000b7da40, 0xc0018607e0, 0x6a, 0xc001081c40, 0x15, 0xc00137d6e0, 0xc0019a4d50, 0xc001238401, 0x0, 0x0, ...)
	/Users/alex/go/src/github.com/pulumi/pulumi/pkg/resource/plugin/provider_plugin.go:263 +0x3f2
github.com/pulumi/pulumi/pkg/resource/deploy.(*stepGenerator).diff(0xc000174310, 0xc0018607e0, 0x6a, 0xc001081c40, 0x15, 0xc00137d530, 0xc00137d6e0, 0xc0019a4d50, 0x1bf2d00, 0xc000b7da40, ...)
	/Users/alex/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/step_generator.go:591 +0x121
github.com/pulumi/pulumi/pkg/resource/deploy.(*stepGenerator).GenerateSteps(0xc000174310, 0x72c02a8, 0xc0019ba180, 0x0, 0x0, 0x0, 0xc0017a5c00)
	/Users/alex/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/step_generator.go:278 +0x1dbf
github.com/pulumi/pulumi/pkg/resource/deploy.(*planExecutor).handleSingleEvent(0xc0005e51c0, 0x1bdca40, 0xc0019ba180, 0xc0017a5cd8)
	/Users/alex/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/plan_executor.go:211 +0x156
github.com/pulumi/pulumi/pkg/resource/deploy.(*planExecutor).Execute.func3(0xc0005e2720, 0xc0005e51c0, 0xc001613070, 0x1be9ba0, 0xc0005f4380, 0x1be9ba0, 0xc0005f4240, 0x10000, 0x1, 0xc0005d2280)
	/Users/alex/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/plan_executor.go:171 +0x280
github.com/pulumi/pulumi/pkg/resource/deploy.(*planExecutor).Execute(0xc0005e51c0, 0x1be9ba0, 0xc0005f4240, 0x1be60e0, 0xc0005e6120, 0x7fffffff, 0x10000, 0x3030303030303001, 0x0, 0x0)
	/Users/alex/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/plan_executor.go:187 +0x4c4
github.com/pulumi/pulumi/pkg/resource/deploy.(*Plan).Execute(0xc0005d8180, 0x1be9ba0, 0xc0005f4240, 0x1be60e0, 0xc0005e6120, 0x7fffffff, 0x10000, 0xc0003fea01, 0x1370000002f, 0x13700000000)
	/Users/alex/go/src/github.com/pulumi/pulumi/pkg/resource/deploy/plan.go:286 +0x87
github.com/pulumi/pulumi/pkg/engine.(*planResult).Walk.func1(0x1be60e0, 0xc0005e6120, 0xc0005e6090, 0x1be9ba0, 0xc0005f4240, 0x1, 0xc001612fd0, 0xc0005e2240)
	/Users/alex/go/src/github.com/pulumi/pulumi/pkg/engine/plan.go:214 +0xc9
created by github.com/pulumi/pulumi/pkg/engine.(*planResult).Walk
	/Users/alex/go/src/github.com/pulumi/pulumi/pkg/engine/plan.go:206 +0x114

```

cc @swgillespie @pgavlin @lukehoban

--- a/pkg/engine/lifecycle_test.go
+++ b/pkg/engine/lifecycle_test.go
@@ -918,7 +918,7 @@ func TestSingleResourceDiffUnavailable(t *testing.T) {
 				DiffF: func(urn resource.URN, id resource.ID,
 					olds, news resource.PropertyMap) (plugin.DiffResult, error) {
 
-					return plugin.DiffResult{Changes: plugin.DiffUnavailable}, errors.New("diff unavailable")
+					return plugin.DiffResult{}, plugin.DiffUnavailable("diff unavailable")
 				},
 			}, nil
 		}),
@@ -935,6 +935,7 @@ func TestSingleResourceDiffUnavailable(t *testing.T) {
 	p := &TestPlan{
 		Options: UpdateOptions{host: host},
 	}
+	resURN := p.NewURN("pkgA:m:typA", "resA", "")
 
 	// Run the initial update.
 	project := p.GetProject()
@@ -943,7 +944,21 @@ func TestSingleResourceDiffUnavailable(t *testing.T) {
 
 	// Now change the inputs to our resource and run a preview.
 	inputs = resource.PropertyMap{"foo": resource.NewStringProperty("bar")}
-	_, err = TestOp(Update).Run(project, p.GetTarget(snap), p.Options, true, p.BackendClient, nil)
+	_, err = TestOp(Update).Run(project, p.GetTarget(snap), p.Options, true, p.BackendClient,
+		func(_ workspace.Project, _ deploy.Target, _ *Journal, events []Event, err error) error {
+			found := false
+			for _, e := range events {
+				if e.Type == DiagEvent {
+					p := e.Payload.(DiagEventPayload)
+					if p.URN == resURN && p.Severity == diag.Warning && p.Message == "diff unavailable" {
+						found = true
+						break
+					}
+				}
+			}
+			assert.True(t, found)
+			return err
+		})
 	assert.NoError(t, err)
 }
 

--- a/pkg/engine/lifecycle_test.go
+++ b/pkg/engine/lifecycle_test.go
@@ -911,6 +911,42 @@ func TestSingleResourceExplicitProviderDeleteBeforeReplace(t *testing.T) {
 	p.Run(t, snap)
 }
 
+func TestSingleResourceDiffUnavailable(t *testing.T) {
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{
+				DiffF: func(urn resource.URN, id resource.ID,
+					olds, news resource.PropertyMap) (plugin.DiffResult, error) {
+
+					return plugin.DiffResult{Changes: plugin.DiffUnavailable}, errors.New("diff unavailable")
+				},
+			}, nil
+		}),
+	}
+
+	inputs := resource.PropertyMap{}
+	program := deploytest.NewLanguageRuntime(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		_, _, _, err := monitor.RegisterResource("pkgA:m:typA", "resA", true, "", false, nil, "", inputs)
+		assert.NoError(t, err)
+		return nil
+	})
+	host := deploytest.NewPluginHost(nil, nil, program, loaders...)
+
+	p := &TestPlan{
+		Options: UpdateOptions{host: host},
+	}
+
+	// Run the initial update.
+	project := p.GetProject()
+	snap, err := TestOp(Update).Run(project, p.GetTarget(nil), p.Options, false, p.BackendClient, nil)
+	assert.NoError(t, err)
+
+	// Now change the inputs to our resource and run a preview.
+	inputs = resource.PropertyMap{"foo": resource.NewStringProperty("bar")}
+	_, err = TestOp(Update).Run(project, p.GetTarget(snap), p.Options, true, p.BackendClient, nil)
+	assert.NoError(t, err)
+}
+
 func TestDestroyWithPendingDelete(t *testing.T) {
 	loaders := []*deploytest.ProviderLoader{
 		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {

--- a/pkg/resource/deploy/step_generator.go
+++ b/pkg/resource/deploy/step_generator.go
@@ -275,7 +275,14 @@ func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, *re
 			// Determine whether the change resulted in a diff.
 			d, diffErr := sg.diff(urn, old.ID, oldInputs, oldOutputs, inputs, prov, allowUnknowns)
 			if diffErr != nil {
-				return nil, result.FromError(diffErr)
+				// If the plugin indicated that the diff is unavailable, assume that the resource will be updated and
+				// report the message contained in the error.
+				if d.Changes == plugin.DiffUnavailable {
+					d.Changes = plugin.DiffSome
+					sg.plan.ctx.Diag.Warningf(diag.RawMessage(urn, diffErr.Error()))
+				} else {
+					return nil, result.FromError(diffErr)
+				}
 			}
 			diff = d
 		}
@@ -585,7 +592,7 @@ func (sg *stepGenerator) diff(urn resource.URN, id resource.ID, oldInputs, oldOu
 	// provider returns an "unknown" diff result, pretend it returned "diffs exist".
 	diff, err := prov.Diff(urn, id, oldOutputs, newInputs, allowUnknowns)
 	if err != nil {
-		return plugin.DiffResult{}, err
+		return diff, err
 	}
 	if diff.Changes == plugin.DiffUnknown {
 		diff.Changes = plugin.DiffSome

--- a/pkg/resource/deploy/step_generator.go
+++ b/pkg/resource/deploy/step_generator.go
@@ -277,8 +277,8 @@ func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, *re
 			if diffErr != nil {
 				// If the plugin indicated that the diff is unavailable, assume that the resource will be updated and
 				// report the message contained in the error.
-				if d.Changes == plugin.DiffUnavailable {
-					d.Changes = plugin.DiffSome
+				if _, ok := diffErr.(plugin.DiffUnavailableError); ok {
+					d = plugin.DiffResult{Changes: plugin.DiffSome}
 					sg.plan.ctx.Diag.Warningf(diag.RawMessage(urn, diffErr.Error()))
 				} else {
 					return nil, result.FromError(diffErr)

--- a/pkg/resource/plugin/provider.go
+++ b/pkg/resource/plugin/provider.go
@@ -93,6 +93,9 @@ const (
 	DiffNone DiffChanges = 1
 	// DiffSome indicates the provider performed a diff and concluded that an update or replacement is needed.
 	DiffSome DiffChanges = 2
+	// DiffUnavailable indicates that the provider is unable to diff a resource, usually because the provider has
+	// some unknown configuration values.
+	DiffUnavailable DiffChanges = 3
 )
 
 // DiffResult indicates whether an operation should replace or update an existing resource.

--- a/pkg/resource/plugin/provider.go
+++ b/pkg/resource/plugin/provider.go
@@ -93,9 +93,6 @@ const (
 	DiffNone DiffChanges = 1
 	// DiffSome indicates the provider performed a diff and concluded that an update or replacement is needed.
 	DiffSome DiffChanges = 2
-	// DiffUnavailable indicates that the provider is unable to diff a resource, usually because the provider has
-	// some unknown configuration values.
-	DiffUnavailable DiffChanges = 3
 )
 
 // DiffResult indicates whether an operation should replace or update an existing resource.
@@ -109,4 +106,19 @@ type DiffResult struct {
 // Replace returns true if this diff represents a replacement.
 func (r DiffResult) Replace() bool {
 	return len(r.ReplaceKeys) > 0
+}
+
+// DiffUnavailableError may be returned by a provider if the provider is unable to diff a resource.
+type DiffUnavailableError struct {
+	reason string
+}
+
+// DiffUnavailable creates a new DiffUnavailableError with the given message.
+func DiffUnavailable(reason string) DiffUnavailableError {
+	return DiffUnavailableError{reason: reason}
+}
+
+// Error returns the error message for this DiffUnavailableError.
+func (e DiffUnavailableError) Error() string {
+	return e.reason
 }

--- a/pkg/resource/plugin/provider_plugin.go
+++ b/pkg/resource/plugin/provider_plugin.go
@@ -262,7 +262,7 @@ func (p *provider) Diff(urn resource.URN, id resource.ID,
 	// Instead, indicate that the diff is unavailable and write a message
 	if !p.cfgknown {
 		logging.V(7).Infof("%s: cannot diff due to unknown config", label)
-		const message = "The provider for this resource has inputs that are not known during preview." +
+		const message = "The provider for this resource has inputs that are not known during preview.\n" +
 			"This preview may not correctly represent the changes that will be applied during an update."
 		return DiffResult{}, DiffUnavailable(message)
 	}

--- a/pkg/resource/plugin/provider_plugin.go
+++ b/pkg/resource/plugin/provider_plugin.go
@@ -257,10 +257,14 @@ func (p *provider) Diff(urn resource.URN, id resource.ID,
 		return DiffResult{}, err
 	}
 
-	// If this function is called, we must have complete configuration for the underlying provider. Per DiffConfig,
-	// any unknown input will cause the provider to be replaced, which will cause all of its resources to be replaced,
-	// and we do not call `Diff` for resources that are being replaced due to a change to their provider reference.
-	contract.Assert(p.cfgknown)
+	// If the configuration for this provider was not fully known--e.g. if we are doing a preview and some input
+	// property was sourced from another resource's output properties--don't call into the underlying provider.
+	// Instead, indicate that the diff is unavailable and write a message
+	if !p.cfgknown {
+		logging.V(7).Infof("%s: cannot diff due to unknown config", label)
+		err := errors.New("cannot diff: this resource's provider has unknown configuration values")
+		return DiffResult{Changes: DiffUnavailable}, err
+	}
 
 	molds, err := MarshalProperties(olds, MarshalOptions{
 		Label: fmt.Sprintf("%s.olds", label), ElideAssetContents: true, KeepUnknowns: allowUnknowns})

--- a/pkg/resource/plugin/provider_plugin.go
+++ b/pkg/resource/plugin/provider_plugin.go
@@ -262,8 +262,9 @@ func (p *provider) Diff(urn resource.URN, id resource.ID,
 	// Instead, indicate that the diff is unavailable and write a message
 	if !p.cfgknown {
 		logging.V(7).Infof("%s: cannot diff due to unknown config", label)
-		err := errors.New("cannot diff: this resource's provider has unknown configuration values")
-		return DiffResult{Changes: DiffUnavailable}, err
+		const message = "The provider for this resource has inputs that are not known during preview." +
+			"This preview may not correctly represent the changes that will be applied during an update."
+		return DiffResult{}, DiffUnavailable(message)
 	}
 
 	molds, err := MarshalProperties(olds, MarshalOptions{

